### PR TITLE
CORE-592: rspace: restructure to support multiple impls.

### PR DESCRIPTION
**Pull request by henrytill**
## Overview

To run RSpace in "replay" mode, we will need to support multiple implementations with different semantics for produce and consume.

This commit moves the core functions into a class called `RSpace` which implements the `ISpace `interface.  An instance of `PureSpace`, which contains the "pure" API, can be constructed from anything that implements the `ISpace` interface.

In the context of "replay" mode, we will construct the Reducer with a `PureSpace` that wraps an (as of yet unwritten) `ReplayRSpace`.

### Does this PR relate to an RChain JIRA issue? 
https://rchain.atlassian.net/browse/CORE-592

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [ ] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
N/A

### CODE REVIEW

**Comment by goral09** [comment]
> To run RSpace in "replay" mode, we will need to support multiple implementations with different semantics for produce and consume.

@henrytill What does _different semantics for produce and consume_ mean in this context?

File: rspace/src/main/scala/coop/rchain/rspace/pure/PureSpace.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package coop.rchain.rspace.pure
+
+import coop.rchain.catscontrib.Capture
+import coop.rchain.rspace._
+
+import scala.collection.immutable.Seq
+
+class PureSpace[F[_], C, P, A, K](space: ISpace[C, P, A, K]) {
```
Comment (goral09) [inline rspace/src/main/scala/coop/rchain/rspace/pure/PureSpace.scala]: If there's going to be a `ReplayRSpace` shouldn't we try to keep naming consistent and call this one `PureRSpace` ?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 2]
```diff
@@ -64,59 +66,63 @@ object SystemProcesses {
...
   //TODO(mateusz.gorski): we decided to look into delivering secp256k1 library (https://github.com/bitcoin-core/secp256k1)
   // as separate jar in the future
-//  def secp256k1Verify(store: IStore[Channel, BindPattern, Seq[Channel], TaggedContinuation],
+//  def secp256k1Verify(space: ISpace[Channel, BindPattern, Seq[Channel], TaggedContinuation],
 //                      dispatcher: Dispatch[Task, Seq[Channel], TaggedContinuation])
 //    : Seq[Seq[Channel]] => Task[Unit] = {
 //    case Seq(Seq(IsByteArray(data), IsByteArray(signature), IsByteArray(pub), ack)) =>
 //      Task.fromTry(Try(Secp256k1.verify(data, signature, pub))).flatMap { verified =>
-//        produce(store, ack, Seq(Channel(Quote(Expr(GBool(verified))))), false)
+//        space.produce(ack, Seq(Channel(Quote(Expr(GBool(verified))))), false)
```
[the remark this answers is not in the document]
Comment (henrytill) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: 😉

File: rspace/src/main/scala/coop/rchain/rspace/pure/PureSpace.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    c.capture(space.install(channels, patterns, continuation))
+
+  def produce(channel: C, data: A, persist: Boolean)(implicit
+                                                     m: Match[P, A],
+                                                     c: Capture[F]): F[Option[(K, Seq[A])]] =
+    c.capture(space.produce(channel, data, persist))
+
+  def getCheckpoint()(c: Capture[F]): F[Blake2b256Hash] = c.capture(space.getCheckpoint())
+
+  def reset(hash: Blake2b256Hash)(c: Capture[F]): F[Unit] = c.capture(space.getCheckpoint())
```
[the remark this answers is not in the document]
Comment (henrytill) [reply rspace/src/main/scala/coop/rchain/rspace/pure/PureSpace.scala]: No: `reset` here is like `git reset` in that it needs an argument to specify which root hash we moving to.

**Comment by henrytill** [comment]
@goral09 

> What does _different semantics for produce and consume_ mean in this context?

In "replay" mode, we will be providing a trace log when instantiating `ReplayRSpace`.  This will "rig" `produce` and `consume` to deterministically choose race winners based on the log.  This allows us to deterministically execute a deployment in the face of the inherent non-determinsm in the way Rholang calls RSpace and the way that RSpace works.

File: rspace/src/main/scala/coop/rchain/rspace/pure/PureSpace.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (goral09) [reply rspace/src/main/scala/coop/rchain/rspace/pure/PureSpace.scala]: but you're calling `space.getCheckpoint()`

**Pull request merged**